### PR TITLE
Serial ID

### PR DESCRIPTION
**Pull request by Yatekii**

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -174,6 +181,12 @@ fn main_try() -> Result<(), failure::Error> {
 
     logging::init(opt.log);
 
+    // If someone wants to list the connected probes, just do that and exit.
+    if opt.list_probes {
+        list_connected_devices()?;
+        std::process::exit(0);
```
Comment (Tiwalun) [inline src/main.rs:187]: I would prefer a return here.

File: src/main.rs [hunk 2]
```diff
@@ -250,23 +263,24 @@ fn main_try() -> Result<(), failure::Error> {
...
-        })?,
+    // If we got a probe selector as an argument, open the probe matching the selector if possible.
+    let mut probe = match opt.probe_selector {
+        Some(selector) => Probe::open(selector)?,
         None => {
             // Only automatically select a probe if there is only
             // a single probe detected.
             if list.len() > 1 {
-                return Err(format_err!("More than a single probe detected. Use the --probe-index argument to select which probe to use."));
+                return Err(format_err!("More than a single probe detected. Use the --probe-selector argument to select which probe to use."));
```
Comment (Tiwalun) [inline src/main.rs:273]: I would suggest naming this just `probe`. probe-selector sounds a bit weird to me.

File: src/main.rs [hunk 3]
```diff
@@ -653,3 +668,20 @@ fn remove_arguments_test() {
 
     assert!(arguments.len() == 0);
 }
+
+/// Lists all connected devices
+fn list_connected_devices() -> Result<(), DebugProbeError> {
+    let links = Probe::list_all();
```
Comment (Tiwalun) [inline src/main.rs:674]: Why is this called links?

File: src/main.rs [hunk 4]
```diff
@@ -35,13 +35,18 @@ struct Opt {
...
+    list_probes: bool,
     #[structopt(name = "disable-progressbars", long = "disable-progressbars")]
     disable_progressbars: bool,
     #[structopt(name = "protocol", long = "protocol", default_value = "swd")]
     protocol: WireProtocol,
-    /// The number associated with the debug probe to use
-    #[structopt(long = "probe-index")]
-    n: Option<usize>,
+    #[structopt(long = "probe-selector")]
+    probe_selector: Option<DebugProbeSelector>,
```
Comment (Tiwalun) [inline src/main.rs:49]: I think we should add some help text here, to explain how the selection works.

File: src/main.rs (hunk 3 above)
Comment (Yatekii) [reply src/main.rs:674]: Somebody named it like this in the probe-rs-cli ;)

File: CHANGELOG.md [hunk 5]
```diff
@@ -7,6 +7,19 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+- Added `Cargo.toml` metadata parsing for specifying the chip (see https://github.com/probe-rs/cargo-flash/pull/31).
+
+### Changed
+
+- Improved error logging by a large marging!
```
Comment (therealprof) [inline CHANGELOG.md:14]:
```suggestion
- Improved error logging by a large margin!
```

**Pull request merged**